cpp/src/strings/combine/concatenate.cu

```diff
@@ -0,0 +1,294 @@
+  offset_type* d_offsets{};
+  char* d_chars{};
+
+  __device__ void operator()(size_type idx)
+  {
+    bool const null_element =
+      thrust::any_of(thrust::seq, d_table.begin(), d_table.end(), [idx](auto const& col) {
+        return col.is_null(idx);
+      });
+    // handle a null row
+    if (null_element && !d_narep.is_valid()) {
+      if (!d_chars) d_offsets[idx] = 0;
+      return;
+    }
+
+    char* d_buffer  = d_chars ? d_chars + d_offsets[idx] : nullptr;
+    size_type bytes = 0;
+    for (auto itr = d_table.begin(); itr < d_table.end(); ++itr) {
+      auto const d_column = *itr;
+      auto const d_str =
+        d_column.is_null(idx) ? d_narep.value() : d_column.element<string_view>(idx);
+      if (d_buffer) d_buffer = detail::copy_string(d_buffer, d_str);
+      bytes += d_str.size_bytes();
+      // separator goes only in between elements
+      if (itr + 1 < d_table.end()) {
+        if (d_buffer) d_buffer = detail::copy_string(d_buffer, d_separator);
+        bytes += d_separator.size_bytes();
+      }
+    }
+    if (!d_chars) d_offsets[idx] = bytes;
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<column> concatenate(table_view const& strings_columns,
+                                    string_scalar const& separator,
+                                    string_scalar const& narep,
+                                    rmm::cuda_stream_view stream,
+                                    rmm::mr::device_memory_resource* mr)
+{
+  auto const num_columns = strings_columns.num_columns();
+  CUDF_EXPECTS(num_columns > 0, "At least one column must be specified");
+  // check all columns are of type string
+  CUDF_EXPECTS(std::all_of(strings_columns.begin(),
+                           strings_columns.end(),
+                           [](auto c) { return c.type().id() == type_id::STRING; }),
+               "All columns must be of type string");
+  if (num_columns == 1)  // single strings column returns a copy
+    return std::make_unique<column>(*(strings_columns.begin()), stream, mr);
+  auto const strings_count = strings_columns.num_rows();
+  if (strings_count == 0)  // empty begets empty
+    return detail::make_empty_strings_column(stream, mr);
+
+  CUDF_EXPECTS(separator.is_valid(), "Parameter separator must be a valid string_scalar");
+  string_view d_separator(separator.data(), separator.size());
+  auto d_narep = get_scalar_device_view(const_cast<string_scalar&>(narep));
+
+  // Create device views from the strings columns.
+  auto d_table = table_device_view::create(strings_columns, stream);
+  concat_strings_fn fn{*d_table, d_separator, d_narep};
+  auto children = make_strings_children(fn, strings_count, 0, stream, mr);
+
+  // create resulting null mask
+  auto valid_mask = cudf::detail::valid_if(
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(strings_count),
+    [d_table = *d_table, d_narep] __device__(size_type idx) {
+      bool null_element = thrust::any_of(
```

Comment:
As discuss before, we are avoiding thrust::any_of/all_of etc... and switch to use thrust::count_if() > 0as thrust logical functions have performance problem and are much slower.

Reply:
I don't think this applies for `thrust::seq`. I did a quick test and found that `count_if` can be slower than `all_of` since it can return early when one non-null element is found. Otherwise, they have equivalent performance.
Given that, I think `all_of` more correctly conveys the intention here.

Reply:
@jrhemstad Probably the issue has been fixed in thrust side (https://github.com/NVIDIA/thrust/issues/1016), can you check that by any chance please?